src/ipa/rpi/controller/rpi/sync.cpp

```diff
@@ -119,100 +144,198 @@ void Sync::initialise()
+
+		/* Approximates frame duration */
+		static int64_t modeled_wall_clock_value = 0;
+		static int64_t last_wall_clock_value = 0;
+		static std::chrono::microseconds delta_mod = 0us;
+		static std::chrono::microseconds lastPayloadFrameDuration = 0us;
+		static std::chrono::microseconds expected = 0us;
+
+		if (!initialise){
+			initialise = true;
+			Sync::initialise_socket();
+		}
 
 		while (true) {
-			int64_t lastWallClock = lastPayload_.nextWallClock;
 			int ret = recvfrom(socket_, &lastPayload_, sizeof(lastPayload_), 0, (struct sockaddr *)&addr_, &addrlen);
-
+			
 			if (ret > 0) {
-				int jitter = lastPayload_.wallClock - lastWallClock;
-				LOG(RPiSync, Info) << "Receive message: seq " << lastPayload_.sequence << " ts " << lastPayload_.wallClock
-						<< " server jitter " << jitter << "us"
-						<< " : next seq " << lastPayload_.nextSequence << " ts " << lastPayload_.nextWallClock
-						<< " est duration " << (lastPayload_.nextWallClock - lastPayload_.wallClock) * 1us / (lastPayload_.nextSequence - lastPayload_.sequence)
-						<< " : readyFrame " << lastPayload_.readyFrame;
+				
+				if (!IPCheck) {
+					char srcIP[INET_ADDRSTRLEN];
+					inet_ntop(AF_INET, &(addr_.sin_addr), srcIP, INET_ADDRSTRLEN);
+					std::string serverIP = srcIP;
+
+					static std::string clientIP = local_adress_IP();
+
+					if (serverIP == clientIP) {
+						usingWallClock_ = "no";
+						LOG(RPiSync, Info) << "Using server time stamp ";
+					} else {
+						LOG(RPiSync, Info) << "Using modelled wall clock ";
+					}
+					LOG(RPiSync, Info) <<"Sever ip: " << serverIP << " client ip: " <<clientIP; 
+					IPCheck = true;
+				}
+
+				if (!syncReady_) {
 					state_ = State::Correcting;
-					frames = 0;						
-
-					if (!syncReady_)
-						readyCountdown_ = lastPayload_.readyFrame + frameCount_;
-
-				} else
-					break;
+				}
+
+				frames = 0;	
+	
+				if (usingWallClock_ == "yes") {
+					modeled_wall_clock_value = trendingClock.modeled_wall_clock(local.wallClock, local.sensorTimestamp, local.sequence);
+					last_wall_clock_value = lastPayload_.wallClock;
+				} else {
+					modeled_wall_clock_value = (local.sensorTimestamp) / 1000;
+					last_wall_clock_value = (lastPayload_.sensorTimestamp) / 1000;
+				}
+
+				lastPayloadFrameDuration = (lastPayload_.nextWallClock - lastPayload_.wallClock) * 1us / (lastPayload_.nextSequence - lastPayload_.sequence);
+				std::chrono::microseconds delta = (modeled_wall_clock_value * 1us) - (last_wall_clock_value * 1us);
+				unsigned int mul = (delta + lastPayloadFrameDuration / 2) / lastPayloadFrameDuration;
+				
+				delta_mod = delta - mul * lastPayloadFrameDuration;				
+
+				if (!syncReady_)
+					readyCountdown_ = lastPayload_.readyFrame + frameCount_;
+				
+				if (!syncReady_ && lastPayload_.readyFrame) {
+					expected = lastPayload_.wallClock * 1us + lastPayload_.readyFrame * lastPayloadFrameDuration;
+				}
+			} else {
+				
+				break;}
+		}
+		if (syncReady_ && !frames) {
```

Comment:
I wonder a little bit about this `frames` variable now. We don't use it much any more, so it seems to be more of a "is this my first time in here" kind of flag.  I wonder if there's some other way to know this, in which case it could just be removed.